File: examples/nirfmx/specan/c/RFmxSpecAnAcpDynamicRange/RFmxSpecAnAcpDynamicRange.c
//Steps:
//1. Open a new RFmx session
//2. Configure Selected Ports
//3. Configure Frequency Reference
//4. Configure the basic signal properties  (Center Frequency, Reference Level and External Attenuation)
//5. Configure RF Attenuation
//6. Select ACP measurement and enable the traces
//7. Configure ACP Measurement Method, Power Units and Averaging Parameters
//8. Configure ACP RBW filter
//9. Configure ACP Sweep Time
//10. Configure ACP Noise Compensation
//11. Configure ACP Carrier Channel Settings (Integration BW, RRC Filter)
//12. Configure ACP Number of Offset Channels
//13. Configure ACP Offset Channel Settings (Offset Frequencies, Integration BW, RRC Filter)
//Use "offset:all" selector string to set a parameter for all the Offset Channels
//14. Initiate Measurement
//15. Fetch ACP Measurements and Traces
//16. Close the RFmx Session

// Include files

#include <stdio.h>
#include <conio.h>
#include <stdlib.h>
#include "niRFmxSpecAn.h"

/* Maximum size of an error message */
#define MAX_ERROR_DESCRIPTION       4096

#define NUMBER_OF_OFFSETS           2
#define AUTO_LEVEL_OFF              0
#define AUTO_LEVEL_ON               1

/* Datatype for input Values */
typedef struct {
   float64 integrationBandwidth;
   int32 RRCFilterEnabled;
   float64 RRCFilterAlpha;
}carrierChannel_t;

typedef struct {
   float64 integrationBandwidth;
   float64 frequencyOffset[NUMBER_OF_OFFSETS];
   int32 RRCFilterEnabled;
   float64 RRCFilterAlpha;
}offsetChannel_t;

int main(int argc, char *argv[])
{
   int i = 0;
   niRFmxInstrHandle instrumentHandle = NULL;

   char errorMessage[MAX_ERROR_DESCRIPTION];
   int32 error = 0, lastErrorCode = 0;

   char *resourceName = "RFSA";
   char *selectedPorts = "";
   float64 centerFrequency = 1e+9;           /* Hz */
   float64 autoSetReferenceLevel;
   float64 referenceLevel = 0.00;            /* dBm */
   float64 externalAttenuation = 0.00;       /* dB */
   float64 integrationBandwidth = 1e+6;
   float64 measurementInterval = 10e-3;
   int32 autoLevel = AUTO_LEVEL_ON;

   int32 rfAttenuationAuto = RFMXSPECAN_VAL_TRUE;
   float64 rfAttenuation = 10.00;            /* dB */

   char *frequencySource = RFMXINSTR_VAL_ONBOARD_CLOCK_STR;
   float64 frequency = 10.0e+6;              /* Hz */

   int32 powerUnits = RFMXSPECAN_VAL_ACP_POWER_UNITS_DBM;
   int32 measurementMethod = RFMXSPECAN_VAL_ACP_MEASUREMENT_METHOD_DYNAMIC_RANGE;
   int32 noiseCompensationEnabled = RFMXSPECAN_VAL_ACP_NOISE_COMPENSATION_ENABLED_TRUE;

   /* Sweep Time */
   int32 sweepTimeAuto = RFMXSPECAN_VAL_ACP_SWEEP_TIME_AUTO_TRUE;
   float64 sweepTimeInterval = 1.00e-3;      /* seconds */

   /* RBW Filter */
   int32 RBWFilterType = RFMXSPECAN_VAL_ACP_RBW_FILTER_TYPE_GAUSSIAN;
   int32 RBWAuto = RFMXSPECAN_VAL_ACP_RBW_AUTO_TRUE;
   float64 RBW = 10.0e+3;                    /* Hz */

   /* Averaging */
   int32 averagingEnabled = RFMXSPECAN_VAL_ACP_AVERAGING_ENABLED_FALSE;
   int32 averagingCount = 10;
   int32 averagingType = RFMXSPECAN_VAL_ACP_AVERAGING_TYPE_RMS;

   float64 timeout = 10.0;                   /* seconds */
   int32 numberOfOffsets = NUMBER_OF_OFFSETS;

   /* Variables to store the results */
   float64 absolutePower;
   int32 offsetMeasArraySize;
   float64 *lowerRelativePower = (float64 *)NULL;
   float64 *upperRelativePower = (float64 *)NULL;
   float64 *lowerAbsolutePower = (float64 *)NULL;
   float64 *upperAbsolutePower = (float64 *)NULL;
   float64 x0 = 0.0, dx = 0.0;
   float32 *spectrum = (float32 *)NULL;
   int32 actualArraySize;

   /* Carrier Channels */
   carrierChannel_t carrierChannelInput;
   offsetChannel_t offsetChannelInput;

   carrierChannelInput.integrationBandwidth = 1.0e+6;    /* Hz */
   carrierChannelInput.RRCFilterEnabled = RFMXSPECAN_VAL_ACP_RRC_FILTER_ENABLED_FALSE;
   carrierChannelInput.RRCFilterAlpha = 0.220;

   /* Offset Channels	*/
   offsetChannelInput.integrationBandwidth = 1.0e+6;     /* Hz */
   offsetChannelInput.RRCFilterEnabled = RFMXSPECAN_VAL_ACP_RRC_FILTER_ENABLED_FALSE;
   offsetChannelInput.RRCFilterAlpha = 0.220;

   offsetChannelInput.frequencyOffset[0] = 1.0e+6;       /* Hz */
   offsetChannelInput.frequencyOffset[1] = 2.0e+6;       /* Hz */

   /* Create new RFmx session */

   RFmxCheckWarn(RFmxSpecAn_Initialize(resourceName, "", &instrumentHandle, NULL));

   /* Configure ACP measurements */
   RFmxCheckWarn(RFmxSpecAn_CfgFrequencyReference(instrumentHandle, "", frequencySource, frequency));
   RFmxCheckWarn(RFmxSpecAn_SetSelectedPorts(instrumentHandle, "", selectedPorts));
   RFmxCheckWarn(RFmxSpecAn_CfgFrequency(instrumentHandle, "", centerFrequency));
   RFmxCheckWarn(RFmxSpecAn_CfgExternalAttenuation(instrumentHandle, "", externalAttenuation));
   RFmxCheckWarn(RFmxSpecAn_CfgRFAttenuation(instrumentHandle, "", rfAttenuationAuto, rfAttenuation));

   if (autoLevel)
   {
      RFmxCheckWarn(RFmxSpecAn_AutoLevel(instrumentHandle, "", integrationBandwidth, measurementInterval, &autoSetReferenceLevel));
      printf("Reference level : %f\n", autoSetReferenceLevel);
   }
   else
   {
      RFmxCheckWarn(RFmxSpecAn_CfgReferenceLevel(instrumentHandle, "", referenceLevel));
   }
   RFmxCheckWarn(RFmxSpecAn_SelectMeasurements(instrumentHandle, "", RFMXSPECAN_VAL_ACP,
      RFMXSPECAN_VAL_TRUE));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgMeasurementMethod(instrumentHandle, "", measurementMethod));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgPowerUnits(instrumentHandle, "", powerUnits));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgAveraging(instrumentHandle, "", averagingEnabled, averagingCount,
      averagingType));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgRBWFilter(instrumentHandle, "", RBWAuto, RBW, RBWFilterType));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgSweepTime(instrumentHandle, "", sweepTimeAuto, sweepTimeInterval));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgNoiseCompensationEnabled(instrumentHandle, "", noiseCompensationEnabled));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgCarrierIntegrationBandwidth(instrumentHandle, "",
      carrierChannelInput.integrationBandwidth));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgCarrierRRCFilter(instrumentHandle, "",
      carrierChannelInput.RRCFilterEnabled,
      carrierChannelInput.RRCFilterAlpha));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgNumberOfOffsets(instrumentHandle, "", numberOfOffsets));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgOffsetArray(instrumentHandle, "",
      offsetChannelInput.frequencyOffset, NULL, NULL,
      numberOfOffsets));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgOffsetIntegrationBandwidth(instrumentHandle, "offset::all",
      offsetChannelInput.integrationBandwidth));
   RFmxCheckWarn(RFmxSpecAn_ACPCfgOffsetRRCFilter(instrumentHandle, "offset::all",
      offsetChannelInput.RRCFilterEnabled,
      offsetChannelInput.RRCFilterAlpha));
   RFmxCheckWarn(RFmxSpecAn_Initiate(instrumentHandle, "", ""));

   /* Retrieve results */
   /* Get actual size of the array */
   RFmxCheckWarn(RFmxSpecAn_ACPFetchOffsetMeasurementArray(instrumentHandle, "", timeout,
      NULL, NULL, NULL, NULL, 0,
      &offsetMeasArraySize));
   if (offsetMeasArraySize > 0)
   {
      lowerRelativePower = (float64 *)malloc(sizeof(float64) * offsetMeasArraySize);
      upperRelativePower = (float64 *)malloc(sizeof(float64) * offsetMeasArraySize);
      lowerAbsolutePower = (float64 *)malloc(sizeof(float64) * offsetMeasArraySize);
      upperAbsolutePower = (float64 *)malloc(sizeof(float64) * offsetMeasArraySize);
      /* Fetch the measurements array */
      RFmxCheckWarn(RFmxSpecAn_ACPFetchOffsetMeasurementArray(instrumentHandle, "", timeout,
         lowerRelativePower, upperRelativePower,
         lowerAbsolutePower, upperAbsolutePower,
         offsetMeasArraySize, NULL));
   }
   RFmxCheckWarn(RFmxSpecAn_ACPFetchCarrierMeasurement(instrumentHandle, "", timeout,
      &absolutePower, NULL, NULL, NULL));
   actualArraySize = 0; x0 = 0.0; dx = 0.0;
   RFmxCheckWarn(RFmxSpecAn_ACPFetchSpectrum(instrumentHandle, "", timeout, NULL, NULL, NULL, 0,
      &actualArraySize));
   if (actualArraySize > 0)
   {
      spectrum = (float32 *)malloc(sizeof(float32) * actualArraySize);
      if (spectrum)
      {
         RFmxCheckWarn(RFmxSpecAn_ACPFetchSpectrum(instrumentHandle, "", timeout, &x0, &dx, spectrum,
            actualArraySize, NULL));
      }
      else
      {
         printf("malloc failed.\n");
         goto Error;
      }
   }

   printf("Carrier Measurements: \n");
   printf("Absolute Power (dBm or dBm/Hz) %f\n", absolutePower);
   printf("---------------------------------------------------\n");


   printf("Offset Channel Measurements: \n");
   for (i = 0; i < offsetMeasArraySize; i++)
   {
      printf("Offset  :  %d\n", i);
      printf("Lower Relative Power (dB)            %f\n", lowerRelativePower[i]);
      printf("Upper Relative Power (dB)            %f\n", upperRelativePower[i]);
      printf("Lower Absolute Power (dBm or dBm/Hz) %f\n", lowerAbsolutePower[i]);
      printf("Upper Absolute Power (dBm or dBm/Hz) %f\n", upperAbsolutePower[i]);
      printf("-------------------------------------------------\n");
   }

Error:
   if (error)
   {
      RFmxSpecAn_GetError(instrumentHandle, &lastErrorCode, MAX_ERROR_DESCRIPTION, errorMessage);
      if (error < 0)
         printf("ERROR: %s\n", errorMessage);
      else
         printf("WARNING: %s\n", errorMessage);
   }
   if (instrumentHandle)
   {
      RFmxSpecAn_Close(instrumentHandle, RFMXSPECAN_VAL_FALSE);
   }
   /* Free allocated memory */
   if (spectrum)
      free(spectrum);
   if (lowerRelativePower)
      free(lowerRelativePower);
   if (upperRelativePower)
      free(upperRelativePower);
   if (upperAbsolutePower)
      free(upperAbsolutePower);
   if (lowerAbsolutePower)
      free(lowerAbsolutePower);
   printf("Press any key to exit\n");
   _getch();

   return error;
}

File: examples/nirfmx/specan/csharp/VS2010/RFmxSpecAnAcpDynamicRange/vb/RFmxSpecAnAcpDynamicRange.vb
'Steps:
'1. Open a new RFmx session
'2. Configure Selected Ports
'3. Configure Frequency Reference
'4. Configure the basic signal properties  (Center Frequency, Reference Level and External Attenuation)
'5. Configure RF Attenuation
'6. Select ACP measurement and enable the traces
'7. Configure ACP Measurement Method, Power Units and Averaging Parameters
'8. Configure ACP RBW filter
'9. Configure ACP Sweep Time
'10. Configure ACP Noise Compensation
'11. Configure ACP Carrier Channel Settings (Integration BW, RRC Filter)
'12. Configure ACP Number of Offset Channels
'13. Configure ACP Offset Channel Settings (Offset Frequencies, Integration BW, RRC Filter)
'Use "offset:all" selector string to set a parameter for all the Offset Channels
'14. Initiate Measurement
'15. Fetch ACP Measurements and Traces
'16. Close the RFmx Session

Imports NationalInstruments.RFmx.InstrMX
Imports NationalInstruments.RFmx.SpecAnMX

Public Class RFmxSpecAnAcpDynamicRangeExample
   Private instrSession As RFmxInstrMX
   Private specAn As RFmxSpecAnMX
   Private resourceName As String

   Const NumberOfOffsets As Integer = 2

   Private selectedPorts As String
   Private centerFrequency As Double, referenceLevel As Double, autoSetReferenceLevel As Double, externalAttenuation As Double, rfAttenuation As Double, rbw As Double,
      frequency As Double, sweepTimeInterval As Double, measurementInterval As Double, timeout As Double
   Private frequencySource As String
   Private autoLevel As Boolean
   Private rfAttenuationAuto As RFmxInstrMXRFAttenuationAuto
   Private powerUnits As RFmxSpecAnMXAcpPowerUnits
   Private measurementMethod As RFmxSpecAnMXAcpMeasurementMethod
   Private noiseCompensationEnabled As RFmxSpecAnMXAcpNoiseCompensationEnabled
   Private rbwFilterType As RFmxSpecAnMXAcpRbwFilterType
   Private rbwAuto As RFmxSpecAnMXAcpRbwAutoBandwidth
   Private sweepTimeAuto As RFmxSpecAnMXAcpSweepTimeAuto
   Private averagingCount As Integer
   Private averagingEnabled As RFmxSpecAnMXAcpAveragingEnabled
   Private averagingType As RFmxSpecAnMXAcpAveragingType

   'Input values
   Private Structure CarrierChannel
      Public integrationBandwidth As Double
      Public rrcFilterEnabled As RFmxSpecAnMXAcpCarrierRrcFilterEnabled
      Public rrcFilterAlpha As Double
   End Structure
   Private carrierChannelInput As CarrierChannel

   Private Structure OffsetChannel
      Public integrationBandwidth As Double
      Public frequencyOffset As Double()
      Public rrcFilterEnabled As RFmxSpecAnMXAcpOffsetRrcFilterEnabled
      Public rrcFilterAlpha As Double
   End Structure
   Private offsetChannelInput As OffsetChannel

   'Output values
   Private absolutePower As Double
   Private lowerRelativePower As Double()
   Private upperRelativePower As Double()
   Private lowerAbsolutePower As Double()
   Private upperAbsolutePower As Double()

   Friend Sub Run()
      Try
         InitializeVariables()
         InitializeInstr()
         ConfigureSpecAn()
         RetrieveResults()
         PrintResults()
      Catch ex As Exception
         DisplayError(ex.Message)
      Finally
         ' Close session 

         CloseSession()
         Console.WriteLine("Press any key to exit.....")
         Console.ReadKey()
      End Try
   End Sub

   Private Sub InitializeVariables()
      ' Initialize input variables 


      resourceName = "RFSA"
      selectedPorts = ""
      centerFrequency = 1000000000.0
      ' Hz 
      referenceLevel = 0.0
      ' dBm 
      externalAttenuation = 0.0
      ' dB 

      rfAttenuationAuto = RFmxInstrMXRFAttenuationAuto.[True]
      rfAttenuation = 10.0
      ' dB 

      frequencySource = RFmxInstrMXConstants.OnboardClock
      frequency = 10000000.0
      ' Hz 

      measurementInterval = 0.01
      ' seconds 
      autoLevel = True

      ' Carrier Channels
      carrierChannelInput.integrationBandwidth = 1000000.0
      carrierChannelInput.rrcFilterEnabled = RFmxSpecAnMXAcpCarrierRrcFilterEnabled.[False]
      carrierChannelInput.rrcFilterAlpha = 0.22

      powerUnits = RFmxSpecAnMXAcpPowerUnits.dBm
      measurementMethod = RFmxSpecAnMXAcpMeasurementMethod.DynamicRange
      noiseCompensationEnabled = RFmxSpecAnMXAcpNoiseCompensationEnabled.[True]

      ' Sweep Time
      sweepTimeAuto = RFmxSpecAnMXAcpSweepTimeAuto.[True]
      sweepTimeInterval = 0.001
      ' seconds 

      ' RBW Filter
      rbwFilterType = RFmxSpecAnMXAcpRbwFilterType.Gaussian
      rbwAuto = RFmxSpecAnMXAcpRbwAutoBandwidth.[True]
      rbw = 10000.0
      ' Hz 

      ' Offset Channels	
      offsetChannelInput.integrationBandwidth = 1000000.0
      ' Hz 
      offsetChannelInput.rrcFilterEnabled = RFmxSpecAnMXAcpOffsetRrcFilterEnabled.[False]
      offsetChannelInput.rrcFilterAlpha = 0.22

      offsetChannelInput.frequencyOffset = New Double(NumberOfOffsets - 1) {}
      offsetChannelInput.frequencyOffset(0) = 1000000.0
      ' Hz 
      offsetChannelInput.frequencyOffset(1) = 2000000.0
      ' Hz 

      'Averaging 
      averagingEnabled = RFmxSpecAnMXAcpAveragingEnabled.[False]
      averagingCount = 10
      averagingType = RFmxSpecAnMXAcpAveragingType.Rms

      timeout = 10.0
      ' seconds 
   End Sub

   Private Sub InitializeInstr()
      ' Create a new RFmx Session 

      instrSession = New RFmxInstrMX(resourceName, "")
   End Sub

   Private Sub ConfigureSpecAn()
      ' Get SpecAn signal 

      specAn = instrSession.GetSpecAnSignalConfiguration()

      ' Configure measurement 

      instrSession.ConfigureFrequencyReference("", frequencySource, frequency)
      specAn.SetSelectedPorts("", selectedPorts)
      specAn.ConfigureFrequency("", centerFrequency)
      specAn.ConfigureExternalAttenuation("", externalAttenuation)
      instrSession.ConfigureRFAttenuation("", rfAttenuationAuto, rfAttenuation)

      If autoLevel Then
         specAn.AutoLevel("", carrierChannelInput.integrationBandwidth, measurementInterval, autoSetReferenceLevel)
         Console.WriteLine("Reference Level(dBm): {0}", autoSetReferenceLevel)
      Else
         specAn.ConfigureReferenceLevel("", referenceLevel)
      End If

      specAn.SelectMeasurements("", RFmxSpecAnMXMeasurementTypes.Acp, True)

      specAn.Acp.Configuration.ConfigureMeasurementMethod("", measurementMethod)
      specAn.Acp.Configuration.ConfigurePowerUnits("", powerUnits)
      specAn.Acp.Configuration.ConfigureAveraging("", averagingEnabled, averagingCount, averagingType)
      specAn.Acp.Configuration.ConfigureRbwFilter("", rbwAuto, rbw, rbwFilterType)
      specAn.Acp.Configuration.ConfigureSweepTime("", sweepTimeAuto, sweepTimeInterval)
      specAn.Acp.Configuration.ConfigureNoiseCompensationEnabled("", noiseCompensationEnabled)
      specAn.Acp.Configuration.ConfigureCarrierIntegrationBandwidth("", carrierChannelInput.integrationBandwidth)
      specAn.Acp.Configuration.ConfigureCarrierRrcFilter("", carrierChannelInput.rrcFilterEnabled, carrierChannelInput.rrcFilterAlpha)

      specAn.Acp.Configuration.ConfigureNumberOfOffsets("", NumberOfOffsets)

      specAn.Acp.Configuration.ConfigureOffsetArray("", offsetChannelInput.frequencyOffset, Nothing, Nothing)

      specAn.Acp.Configuration.ConfigureOffsetIntegrationBandwidth("offset::all", offsetChannelInput.integrationBandwidth)

      specAn.Acp.Configuration.ConfigureOffsetRrcFilter("offset::all", offsetChannelInput.rrcFilterEnabled, offsetChannelInput.rrcFilterAlpha)

      specAn.Initiate("", "")
   End Sub

   Private Sub RetrieveResults()
      ' Retrieve results 

      Dim spectrum As Spectrum(Of Single) = Nothing
      Dim totalRelativePower As Double, carrierFrequency As Double, integrationBandwidth As Double

      specAn.Acp.Results.FetchOffsetMeasurementArray("", timeout, lowerRelativePower, upperRelativePower, lowerAbsolutePower, upperAbsolutePower)

      specAn.Acp.Results.FetchCarrierMeasurement("", timeout, absolutePower, totalRelativePower, carrierFrequency, integrationBandwidth)

      specAn.Acp.Results.FetchSpectrum("", timeout, spectrum)
   End Sub

   Private Sub PrintResults()
      Console.WriteLine("-----------------Carrier Measurements-----------------" & vbLf)
      Console.WriteLine("Absolute Power (dBm or dBm/Hz)       {0}", absolutePower)

      Console.WriteLine(vbLf & "--------------Offset Channel Measurements-------------" & vbLf)
      For i As Integer = 0 To NumberOfOffsets - 1
         Console.WriteLine("----Offset {0}" & vbLf, i)
         Console.WriteLine("Lower Relative Power (dB)            {0}", lowerRelativePower(i))
         Console.WriteLine("Upper Relative Power (dB)            {0}", upperRelativePower(i))
         Console.WriteLine("Lower Absolute Power (dBm or dBm/Hz) {0}", lowerAbsolutePower(i))
         Console.WriteLine("Upper Absolute Power (dBm or dBm/Hz) {0}", upperAbsolutePower(i))
      Next
      Console.WriteLine("-------------------------------------------------" & vbLf)
   End Sub

   Private Sub CloseSession()
      Try
         If specAn IsNot Nothing Then
            specAn.Dispose()
            specAn = Nothing
         End If

         If instrSession IsNot Nothing Then
            instrSession.Close()
            instrSession = Nothing
         End If
      Catch ex As Exception
         DisplayError(ex.Message)
      End Try
   End Sub

   Private Shared Sub DisplayError(message As String)
      Console.WriteLine("ERROR:" & vbLf & message)
   End Sub
End Class
